keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md

```diff
@@ -0,0 +1,902 @@
+in ResourceSlices and allows for more compact device definitions. It also enables
+defining devices with more attributes, capacities and consumed counters than without
+mixins. Mixins can also be used in counter sets.
+
+## Motivation
+
+DRA requires that drivers publish all available devices on a node/cluster in
+`ResourceSlice` objects. There are scenarios where the number of devices
+can be pretty large and each device might have a relatively large amount
+of metadata associated with it, primarily in the form of attributes and
+capacity. This has a few consequences:
+
+* Several of the devices might have similar metadata, resulting in a lot of
+  duplication between the published devices.
+* The size of the data required to specify each device reduces the number of
+  devices that can be defined in a single ResourceSlice without hitting the
+  Kubernetes size limit.
+
+The latter can be addressed by splitting the devices across multiple
+`ResourceSlice`s within a single pool, but that isn't always an option.
+In particular, DRA currently doesn't allow sharing counters across `ResourceSlice`s,
+meaning that the number of devices that can fit into a single `ResourceSlice`
+also limits the number of partitionable devices for a single physical device.
+
+### Goals
+
+- Enable a more compact way to define devices in ResourceSlices so duplication can
+  be reduced and a larger number of devices can be published within a single
+  ResourceSlice.
+- Enable defining devices with more attributes, capacities, and consumed counters.
+- Enable defining counter sets with more counters.
+
+### Non-Goals
+
+- Not part of the plan for alpha: developing kubectl command or plugin to let
+  users see the flattened device definitions. Mixins does make it harder to find
+  the full definition for a specific device, so this might be added to the scope
+  for Beta or GA.
+
+## Proposal
+
+The proposal has two parts to it, the definition of mixins and the
+mechanism for referencing mixins from devices and counter sets.
+
+A new `Mixins` field will be added to the `ResourceSliceSpec` as an
+optional field. It will have three properties, one for each of the three
+types of mixins that will be supported:
+
+1. The `CounterSet` field defines a list of named `CounterSetMixins`.
+   These define counters that can be used to extend the counters
+   explicitly defined in a `CounterSet`. This allows for reduced duplication
+   if there are many identical physical devices that must be represented as
+   `CounterSet`s. `CounterSetMixins` can not be referenced directly by devices.
+
+1. The `Device` field is a list of named `DeviceMixin`s. These define
+   attributes and capacities that can be used to extend what is defined
+   explicitly in `Device`. `DeviceMixin`s cannot be allocated directly, but can
+   only be referenced by devices.
+
+1. The `DeviceCounterConsumption` field defines a list of named
+   `DeviceCounterConsumptionMixin`s. These define counters that can be
```

Comment:
I might tweak this a bit to have something like:

These define the pattern of consumption of counters, distinct from the specific underlying counter set from which they are being consumed.

Rather than "these define counters"

Reply:
Updated